compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -644,7 +644,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
                     diagnostics_builder.note("for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/disjoint-capture-in-closures.html>");
-                    let closure_body_span = self.tcx.hir().span(body_id.hir_id);
+
+                    let mut closure_body_span = self.tcx.hir().span(body_id.hir_id);
+
+                    // If the body was entirely expanded from a macro
+                    // invocation, as the body is not contained inside the
+                    // closure span. In that case, we walk up the expansion
+                    // until we find the span before the expansion.
+                    while !closure_body_span.is_dummy() && !closure_span.contains(closure_body_span) {
```

Comment:
If we end with a dummy span, then the overall diagnostic will end up worse than it currently is. Can you restore `closure_body_span` to its original value if the loop ends up setting it to `DUMMY_SP`?

Reply:
I kept the dummy span for the suggestion part. Otherwise `.span_to_snippet` is going to return the wrong snippet and the suggestion will be wrong and confusing. The fallback span now uses the entire closure, which chas the highest chance of pointing to the place where modifications need to be made.